In [1]:
import numpy as np
import pandas as pd
from collections import Counter
from tqdm import tqdm as tqdm

In [2]:
followers_df = pd.read_csv("/Users/marzie/Documents/Data/followers_df.csv")
followers = followers_df[['user_id', 'podcaster_id']].set_index('user_id')
followers.head()

podcaster_id
user_id               
38719947      21543729
37434280      21543729
22013860      21543729
28885466      21543729
30848189      21543729

In [3]:
all_podcasts_with_channel = pd.read_json("all_podcasts_with_channel.json")
all_podcasts_with_channel.rename(columns= {'user_suid':'podcaster_id'}, inplace = True)
all_podcasts = all_podcasts_with_channel[['cid', 'podcaster_id', 'data']].set_index('cid')
all_podcasts.head()

podcaster_id                                               data
cid                                                                     
2598530    17293748.0  {'uid': '0e242cb9dea94a00aa50efbf010fc0c5', 'o...
4894772    17293748.0  {'uid': '0e242cb9dea94a00aa50efbf010fc0c5', 'e...
1423226     7697322.0  {'uid': 'c4dad8b7604c48f1aecd8f08105dca6f', 'e...
1482589    16024112.0  {'uid': 'd62534a4513e46428378e198ffd1ce9a', 'o...
2356600    16024112.0  {'uid': 'd62534a4513e46428378e198ffd1ce9a', 'e...

In [4]:
my_podcasts = pd.read_csv('my_podcasts_df.csv')
my_cids_list = my_podcasts['cid'].tolist()

In [5]:
my_podcast_with_data = all_podcasts.loc[my_cids_list].reset_index()
my_podcast_with_data

cid  podcaster_id                                               data
0     4894772    17293748.0  {'uid': '0e242cb9dea94a00aa50efbf010fc0c5', 'e...
1     2356600    16024112.0  {'uid': 'd62534a4513e46428378e198ffd1ce9a', 'e...
2     4857685    16682348.0  {'uid': 'b188e784539c480c84c173b29675e858', 'e...
3     2748108    14199420.0  {'uid': 'af088f29b89a422198b86988b2619c4f', 'e...
4     2400145    15814516.0  {'uid': '8097ad5aa05d4e9eba396743f11c658d', 'e...
...       ...           ...                                                ...
3875  5423676    20495886.0  {'uid': '4cc555ff69ed431a94c7b7cffb7fd91d', 'o...
3876  2131745    13501707.0  {'uid': '68d77f432d5d4ae29029f266b3cabd61', 'e...
3877  4711535    24170434.0  {'uid': '18a9ef1a3bf447e7b26f315796842ab3', 'e...
3878  5297748    18840456.0  {'uid': 'c65835ff048448e3820ebd9f1e676fe1', 'o...
3879  3631144    18930934.0  {'uid': 'e61cb2a5e9c34261a17fb808620260d7', 'o...

[3880 rows x 3 columns]

# categories

<div dir=rtl >
<font  size=3>
در گام اول می‌خواهیم بدانیم در کل چند کتگوری متمایز در پادکست‌هایمان وجود دارد. کتگوری هر پادکست را نیز در یک لیست نگه می‌داریم. 

In [6]:
list_of_categories = [[] for i in range(len(my_podcast_with_data))]
all_categories = []

for i in range(len(my_podcast_with_data)):
    category_i = my_podcast_with_data['data'][i]['categories']
    list_of_categories[i] = category_i
    all_categories = all_categories + category_i

In [203]:
#counting the occurrences of the categories
category_dict = dict(Counter(all_categories))

#sort by the frequency of occurrences
keyslist = list(category_dict.keys())
values = list(category_dict.values())
sorted_value_index = np.argsort(values)
sorted_dict = {keyslist[i]: values[i] for i in sorted_value_index}

print('number of unique categories', len(sorted_dict))
print()
print('categories and their frequency of occurrences:')
print()
print(sorted_dict)

number of unique categories 116

categories and their frequency of occurrences:

{10122: 1, 10120: 1, 10121: 1, 10116: 2, 10052: 2, 10111: 2, 10051: 3, 10109: 3, 10055: 3, 10093: 3, 10069: 4, 10065: 4, 10079: 4, 10049: 5, 10050: 5, 10094: 6, 10096: 6, 10089: 7, 10038: 7, 10126: 8, 10128: 8, 10081: 9, 10060: 10, 10034: 10, 10061: 10, 10016: 10, 10064: 12, 10130: 12, 10092: 12, 10085: 12, 10037: 12, 10127: 14, 10077: 14, 10098: 15, 10036: 15, 10097: 15, 10045: 16, 10099: 17, 10047: 18, 10025: 18, 10006: 19, 10078: 20, 10024: 21, 10076: 22, 10090: 25, 10117: 25, 10067: 27, 10112: 27, 10068: 29, 10074: 30, 10053: 30, 10070: 33, 10082: 33, 10054: 33, 10027: 35, 10124: 35, 10073: 36, 10057: 37, 10029: 39, 10056: 40, 10087: 42, 10072: 42, 10042: 45, 10080: 47, 10062: 48, 10106: 49, 10095: 51, 10129: 54, 10031: 60, 10026: 62, 10023: 65, 10105: 66, 10100: 72, 10104: 74, 10102: 74, 10041: 76, 10040: 83, 10107: 85, 10013: 86, 10030: 88, 10005: 96, 10032: 97, 10046: 103, 10033: 115, 10103: 131, 10

<div dir=rtl >
<font  size=3>
برای بررسی کتگوری‌ها پادکست‌هایی با کتگوری مشخص را نشان می‌دهیم.

In [214]:
specific_category = 10071

category = np.where(category_array==specific_category)[0]

#show podcasts with specific category
my_podcasts.loc[category].head(10)

Unnamed: 0      cid  podcaster_id  \
5            12  2480138    12347707.0   
21         1391  4967933    25913181.0   
25         3756  3980827    21638048.0   
33         4000  4565579    22595921.0   
43         2471  4132406    11007494.0   
59         2698  1362236     7521306.0   
77          130  5433705     9917438.0   
93         4008  4796013    21584022.0   
101         328  4852633    11533806.0   
127        4022  5593650    22414402.0   

                                                 title  episode_count  \
5                                  پلی لیست | PlayList            288   
21                                         ترانه گرافی             26   
25               داستان های احسانو / daastaan ehsaanoo             59   
33   هنر ریلکسیشن، موسیقی بیکلام | Art of relaxatio...            174   
43     RJ Podcast | آموزش زبان انگلیسی با فیلم و ترانه            145   
59                   GapDive پادکست موسیقی متن گپ دایو             98   
77                                       Album | آلبوم             51   
93                                                 شعر              5   
101                        AhangeShab | پادکست آهنگ شب            147   
127                                      ♪My Playlist♬             58   

     sub_count  primary_cid                          name  
5       725447          NaN                   مختار رزمجو  
21      160468          NaN   از سری پادکست های کشورگرافی  
25      151369          NaN  احسان عبدی پور ehsanabdipoor  
33      114886          NaN                          Reza  
43       95661          NaN                   Reza Jamali  
59       73680          NaN                      melodive  
77       64327          NaN                   Bardia Barj  
93       56196          NaN                محمد حسین‌نژاد  
101      53961          NaN                  mahdi aghaee  
127      43163          NaN                       VALENSY

<div dir=rtl >
<font  size=3>
با امتحان کردن کتگوری‌های مختلف به نظر می‌رسد داده‌ی نسبتا قابل اعتمادی است.

<div dir=rtl >
<font  size=3>
کاربرانی که کمتر از ۵ پادکستر را دنبال می‌کنند را حذف می‌کنیم.

In [217]:
#find out how many podcasters each user follows
follower_df_user_id_s = followers.index.values.tolist() 

user_sub = dict(Counter(follower_df_user_id_s))

following_df = pd.DataFrame(
    data=list(user_sub.items()),
    columns=['user_id', 'following']).set_index('user_id')

following_df.head()

following
user_id            
38719947         20
37434280         11
22013860        185
28885466         40
30848189         58

In [221]:
mask = following_df['following'] > 4
my_followers = followers.loc[following_df[mask].index.values.tolist()]
my_followers

podcaster_id
user_id               
38719947      21543729
38719947      23250574
38719947      14077619
38719947      20358403
38719947      14078348
...                ...
21290714      19385935
21290714      16045081
21290714      20517596
21290714      15092940
21290714      13855237

[18599616 rows x 1 columns]

## categories function

In [279]:
def categories(user_id):
    
    #five repetitive categories
    remove_categorys = set([10001, 10021, 10014, 10004, 10039])
    
    categories=[]
    
    podcaster_id_s = my_followers.loc[user_id]['podcaster_id'].tolist()
    
    index = my_podcast_with_data[my_podcast_with_data['podcaster_id'].isin(podcaster_id_s)].index.values.tolist() 
    
    for i in range(len(index)):
        categories = categories + list_of_categories[index[i]]
        
    sorted_categories = [key for key, value in Counter(categories).most_common()]
    
    sorted_categories_rm = [i for i in sorted_categories if i not in remove_categorys]
    
    return len(sorted_categories), len(sorted_categories_rm), sorted_categories, sorted_categories_rm

In [280]:
list_of_my_user_ids = np.unique(my_followers.index.values)

In [281]:
number_of_categories = np.zeros(len(list_of_my_user_ids))
number_of_categories_rm = np.zeros(len(list_of_my_user_ids))
sorted_categories = [[] for i in range(len(list_of_my_user_ids))]
sorted_categories_rm = [[] for i in range(len(list_of_my_user_ids))]

for i, user_id in enumerate(tqdm((list_of_my_user_ids))):
    user_categories = categories(user_id)
    number_of_categories[i] = user_categories[0]
    number_of_categories_rm[i] = user_categories[1]
    sorted_categories[i] = user_categories[2]
    sorted_categories_rm[i] = user_categories[3]

100%|██████████████████████████████| 1690671/1690671 [4:38:55<00:00, 101.02it/s]


In [282]:
my_user_categorys = np.zeros((len(list_of_my_user_ids), 5))

In [328]:
for i in range(len(my_user_categorys)):
    
    if number_of_categories_rm[i] >= 5: 
        my_user_categorys[i] = sorted_categories_rm[i][:5]
        
    else:
        my_user_categorys[i, :int(number_of_categories_rm[i])] = list(sorted_categories_rm[i])

<div dir=rtl >
<font  size=3>
با توجه به اینکه تعداد یوزرهایی که کمتر از ۵ کتگوری دارند به نسبت کل کم است، آن‌ها را حذف می‌کنیم.

In [330]:
indices = np.where(number_of_categories_rm < 5)[0]
len(indices)

3978

In [331]:
my_user_categorys_updated = np.delete(my_user_categorys, indices, axis=0)

list_of_my_user_ids_updated = np.delete(list_of_my_user_ids, indices, axis=0)

In [343]:
user_categories_df = pd.DataFrame(
    data=my_user_categorys_updated,
    index=list_of_my_user_ids_updated, 
    columns=['category1', 'category2', 'category3', 'category4', 'category5']
)
user_categories_df.index.name = 'user_id'
user_categories_df

category1  category2  category3  category4  category5
user_id                                                        
1022618     10101.0    10058.0    10022.0    10002.0    10028.0
1032649     10101.0    10091.0    10012.0    10043.0    10104.0
1038639     10101.0    10002.0    10028.0    10022.0    10043.0
1049063     10101.0    10002.0    10028.0    10009.0    10071.0
1049719     10101.0    10009.0    10071.0    10022.0    10103.0
...             ...        ...        ...        ...        ...
39259514    10101.0    10058.0    10048.0    10044.0    10059.0
39259828    10101.0    10058.0    10044.0    10102.0    10009.0
39261381    10022.0    10012.0    10091.0    10101.0    10058.0
39261524    10101.0    10058.0    10048.0    10044.0    10059.0
39261848    10101.0    10059.0    10063.0    10043.0    10058.0

[1686693 rows x 5 columns]

In [344]:
user_categories_df.to_csv("/Users/marzie/Documents/Data/user_categories_df")

In [345]:
my_followers_updatet = my_followers.loc[user_categories_df.index.values.tolist()]

In [346]:
my_followers_updatet

podcaster_id
user_id               
1022618       13480902
1022618       19399631
1022618       15572542
1022618       18150210
1022618       17363840
...                ...
39261848      12347707
39261848      21673341
39261848      24438995
39261848      19624777
39261848      32765155

[18578850 rows x 1 columns]

In [347]:
my_followers_updatet.to_csv("/Users/marzie/Documents/Data/my_followers_df")

In [355]:
print(len(user_categories_df['category1'].unique()))
print()
print(len(user_categories_df['category2'].unique()))
print()
print(len(user_categories_df['category3'].unique()))
print()
print(len(user_categories_df['category4'].unique()))
print()
print(len(user_categories_df['category5'].unique()))

51

78

90

93

101


<div dir=rtl >
<font  size=3>
احتمالا بهتر بود ششمین کتگوری پر تکرار یعنی 10101 را نیز در نظر نمی‌گرفتیم. اگر نتایج غیر منطقی بود به این بخش برمی‌گردیم و این موضوع را اصلاح می‌کنیم. 